subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/normalization/KotlinCompileClasspathFingerprinter.kt

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.kotlin.dsl.normalization
+
+import org.gradle.api.internal.cache.StringInterner
+import org.gradle.api.internal.changedetection.state.AbiExtractingClasspathResourceHasher
+import org.gradle.api.internal.changedetection.state.CachingResourceHasher
+import org.gradle.api.internal.changedetection.state.ResourceSnapshotterCacheService
+import org.gradle.api.internal.changedetection.state.ZipHasher
+import org.gradle.api.tasks.CompileClasspathNormalizer
+import org.gradle.api.tasks.FileNormalizer
+import org.gradle.internal.fingerprint.FileCollectionSnapshotter
+import org.gradle.internal.fingerprint.classpath.CompileClasspathFingerprinter
+import org.gradle.internal.fingerprint.classpath.impl.ClasspathFingerprintingStrategy
+import org.gradle.internal.fingerprint.impl.AbstractFileCollectionFingerprinter
+import org.gradle.internal.snapshot.RegularFileSnapshot
+import org.gradle.kotlin.dsl.support.loggerFor
+
+
+class KotlinCompileClasspathFingerprinter(
+    cacheService: ResourceSnapshotterCacheService,
+    fileCollectionSnapshotter: FileCollectionSnapshotter,
+    stringInterner: StringInterner
+) : AbstractFileCollectionFingerprinter(
+    ClasspathFingerprintingStrategy.compileClasspath(
+        CachingResourceHasher(AbiExtractingClasspathResourceHasher(KotlinApiClassExtractor()), cacheService),
+        cacheService,
+        stringInterner,
+        CompileAvoidanceExceptionReporter()
+    ),
+    fileCollectionSnapshotter
+),
+    CompileClasspathFingerprinter {
+
+    override fun getRegisteredType(): Class<out FileNormalizer> {
+        return CompileClasspathNormalizer::class.java
+    }
+}
+
+
+private
+class CompileAvoidanceExceptionReporter : ZipHasher.HashingExceptionReporter {
+    override fun report(zipFileSnapshot: RegularFileSnapshot, e: Exception) {
+        val jarName = zipFileSnapshot.name
+        if (e is CompileAvoidanceException && !isExternalDependency(jarName)) {
+            logger.warn("Can not use Kotlin build script compile avoidance with {}: {}", jarName, e.message)
+        }
+    }
+
+    private
+    fun isExternalDependency(jarFileName: String) =
+        jarFileName.matches(".*\\d\\.jar".toRegex())
```

Comment:
This is not a good and correct way to detect if a dependency is an external one as opposed to buildSrc/included build jar.

Reply:
If we wanted to only filter out the Gradle related jars, it is possible to filter out Gradle internal dependencies using https://github.com/gradle/gradle/blob/b34e91753ed81aa16360b700e179741da005c780/subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinScriptClassPathProvider.kt#L107-L109 here, however, this does not capture the ones from the `kotlin-dsl` plugin. And I'd say we want to exclude not only Gradle+Kotlin DSL jars from warning that they won't participate in compile avoidance, but rather do it for every jar that comes as an external dependency.

Perhaps the existing naive approach is good enough.

Open to suggestions.

Reply:
I think the easier thing to do for now is to not try to warn the user, make this an info level log and see if there are complaints. The real fix is to make inline methods work properly or narrow their impact more.

So we can just not worry about detecting this at all and stick with just some logging.